### ISSUE TITLE
optimisticUpdater infinite loop

### ISSUE THREAD

**Issue by kaywyeex** [facebook/relay#2600]
In the code below, the optimisticUpdater ends up in an infinite loop.
It only occurs when I set a nested record on a newly created record, and set the linked record on a parent.
eg.
```
1. const newNode = store.create(id, type)
2. newNode.getOrCreateLinkedRecord(key)
3. parent.setLinkedRecords([...oldNodes, newNode])
```

Why does it end up in a loop? How can I prevent it?

### Resolves values and sets them on given record
```
function setValues(node, input) {
  Object.keys(input).forEach(key => {
    const value = input[key];
    if (Array.isArray(value)) node.setLinkedRecords(value, key);
    else if (isPlainObject(value)) {
      Object.keys(value).forEach(nestedKey => {
        const nestedNode = node.getOrCreateLinkedRecord(key);
        setValues(nestedNode, { [nestedKey]: value[nestedKey] });
      });
    } else node.setValue(value, key);
  });
}
```
### Create handler
```
handleCreate = () => {
    const recordUpdater = ({ store, node }) => {
      const parent = store.get(this.props.customer.id);
      const nodes = (parent.getLinkedRecords("customerAddresses") || []).concat(node);

      parent.setLinkedRecords(nodes, "customerAddresses");
    };


    const config = {
      optimisticUpdater: store => {
        const id = `client:newCustomerAddress:${tempId++}`;
        const node = store.create(id, "CustomerAddress");

        node.setValue(id, "id");
        setValues(node, input);

        recordUpdater({ store, node });
      }
    };

    const input = {
      addressType: "Private",
      customerId: this.props.customer.id,
      address: {
        roadName: ""
      }
    };

      commitMutation(this.props.environment, {
        ...config,
        mutation,
        variables: { input },
        onCompleted: (resp, errors) => {
          if (errors) reject(errors);
          else resolve(resp);
        },
        reject
      })
  };
```

**Comment by sibelius** [comment]
It looks like your input is a complex object with cyclic fields

**Comment by kaywyeex** [comment]
I agree that would make sense, however my input is merely like so:
Note that adress {} is what breaks it. 🤔 
```
      address: {
        roadName: ""
      },
      addressType: AddressType.Private,
      customerId: this.props.customer.id

```

**Comment by sibelius** [comment]
can you console.log your input inside setValues

I think the problem is there

**Comment by kaywyeex** [comment]
Here's the result of logging setValues. It seems to make sense.
Also, each iteration of the infinite loop doesn't add a new node to my total array of existing nodes, but it just overwrites the already created one.
eg. ```client:newCustomerAddress:0 => client:newCustomerAddress:1)```

## First iteration
```
function setValues(node, input) {
  console.log("NODE: ", node, "INPUT:", input);
```
... =>
```
NODE:
RelayRecordProxy {_dataID: "client:newCustomerAddress:0", _mutator: RelayRecordSourceMutator, _source: RelayRecordSourceProxy}_dataID: "client:newCustomerAddress:0"_mutator: RelayRecordSourceMutator {_backup: RelayInMemoryRecordSource, _base: RelayInMemoryRecordSource, _sink: RelayInMemoryRecordSource, __sources: Array(2)}_source: RelayRecordSourceProxy {__mutator: RelayRecordSourceMutator, _handlerProvider: ƒ, _proxies: {…}}

INPUT:
proto__: Object {address: {…}, addressType: "Private", customerId: "SUN1c3RvbWVyOjM3"}
```

## Second, recursive iteration
```
NODE:
RelayRecordProxy {_dataID: "client:newCustomerAddress:0:address", _mutator:RelayRecordSourceMutator, _source: RelayRecordSourceProxy}

INPUT:
{roadName: ""}
```

**Comment by sibelius** [comment]
which relay store commands break, and with which params?

**Comment by kaywyeex** [comment]
Not sure I understand the question -- could you rephrase? :-)

**Comment by sibelius** [comment]
`node.setLinkedRecords` breaks? or which command it breaks?

can you also post the full stacktrace?

**Comment by kaywyeex** [comment]
Oh -- that's the problem (kinda). Nothing breaks!
When I run the mutation, relay calls optimisticUpdater infinitely, until I get a ```maximum callstack size exceeded```.

If I console.log(node) inside optimisticUpdater function
```
optimisticUpdater: store => {
        const id = `client:newCustomerAddress:${tempId++}`;
        const node = store.create(id, "CustomerAddress");

        console.log(node)

        node.setValue(id, "id");
        setValues(node, input);

        recordUpdater({ store, node });
      }
```
... you would see
```
RelayRecordProxy {_dataID: "client:newCustomerAddress:0", _mutator: RelayRecordSourceMutator, _source: RelayRecordSourceProxy}
RelayRecordProxy {_dataID: "client:newCustomerAddress:1", _mutator: RelayRecordSourceMutator, _source: RelayRecordSourceProxy}
RelayRecordProxy {_dataID: "client:newCustomerAddress:2", _mutator: RelayRecordSourceMutator, _source: RelayRecordSourceProxy}
RelayRecordProxy {_dataID: "client:newCustomerAddress:3", _mutator: RelayRecordSourceMutator, _source: RelayRecordSourceProxy}
RelayRecordProxy {_dataID: "client:newCustomerAddress:4", _mutator: RelayRecordSourceMutator, _source: RelayRecordSourceProxy}
RelayRecordProxy {_dataID: "client:newCustomerAddress:5", _mutator: RelayRecordSourceMutator, _source: RelayRecordSourceProxy}
RelayRecordProxy {_dataID: "client:newCustomerAddress:6", _mutator: RelayRecordSourceMutator, _source: RelayRecordSourceProxy}

... and so on!
```

**Comment by sibelius** [comment]
the problem is on `setValues`

`const nestedNode = node.getOrCreateLinkedRecord(key);` will return a RecordProxy from relay store that is a complex object with cyclic fields

you should avoid this approach

**Comment by kaywyeex** [comment]
I agree it's causing the problem, but how else would you set a nested object?
Should you just avoid this all together? Seems pretty basic :/

**Comment by kaywyeex** [comment]
Once the response from the server comes, it'll have
```
customerAddress: {
    address: {
        roadName: ""
    }
}
```
And if I don't set address to be an object in the optisticUpdater, it isn't really worth much 🤔

**Comment by sibelius** [comment]
try to use `setLinkedRecords` for objects https://facebook.github.io/relay/docs/en/relay-store.html#setlinkedrecordrecord-recordproxy-name-string-arguments-object

**Comment by kaywyeex** [comment]
You mean for the "address"?
That throws ```Uncaught Error: RelayRecordProxy#setLinkedRecords(): Expected records to be an array, got `[object Object]`.```

**Comment by sibelius** [comment]
what about setValue or copyFieldsFromSource https://facebook.github.io/relay/docs/en/relay-store.html#copyfieldsfromsourcerecord-recordproxy-void?

**Comment by sibelius** [comment]
my bad, try `setLinkedRecord` 

```
setLinkedRecord(
    record: RecordProxy,
    name: string,
    arguments?: ?Object,
  ): RecordProxy;
```

https://facebook.github.io/relay/docs/en/relay-store.html#setlinkedrecordrecord-recordproxy-name-string-arguments-object

I've got the wrong link

**Comment by kaywyeex** [comment]
setValue only works for scalar values(string, int, etc.)

copyFieldsFromSource would require me to copy from somewhere.
Doesn't really make sense :D 

```
        const nestedNode = store.create(node.getDataID() + ":address", "Address");
        node.setLinkedRecord(nestedNode, "address");
```
Gives the same result as ```getOrCreateLinkedRecord```

**Comment by josephsavona** [comment]
As @sibelius noted, `setValues()`is the problem. `setLinkedRecords` accepts array of record proxies, and `setLinkedRecord` accepts a single record proxy. It's possible to write something like your `setValues()` function that works, but it would have to distinguish the types of objects (e.g array of scalars vs array of objects) and convert all nested objects into record proxies with e.g. `store.create` or similar.

**Comment by josephsavona** [comment]
See previous comment for a solution.
